torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -143,48 +138,31 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
 }
 
 void ProcessGroupAgent::enqueue(SendWork work) {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-  sendQueue_.emplace_back(std::move(work));
-  lock.unlock();
-
-  workProduceCV_.notify_one();
-}
-
-// making sure tensors are not deleted before send finishes
-void ProcessGroupAgent::sendLoop() {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-
-  while (!stop_) {
-    if (sendQueue_.empty()) {
-      workProduceCV_.wait(lock);
-      continue;
-    }
-
-    auto work = std::move(sendQueue_.front());
-    sendQueue_.pop_front();
-    lock.unlock();
-
-    workConsumeCV_.notify_one();
-
-
-    std::stringstream ss;
-    serialize(work.message_, ss);
-    std::string str = ss.str();
-
-    std::vector<torch::Tensor> preamble = {
-      torch::tensor(
-        {
-          (int64_t)pg_->getRank(),
-          (int64_t)str.length(),
-        }, {torch::kLong})
-    };
-    pg_->send(preamble, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
-    std::vector<torch::Tensor> payload =
-        {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
-    pg_->send(payload, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
-
-    lock.lock();
-  }
+  sendThreadPool_.run(std::bind(
+    [&](SendWork work) {
+      std::stringstream ss;
+      serialize(work.message_, ss);
+      std::string str = ss.str();
+
+      std::vector<torch::Tensor> preamble = {
+        torch::tensor(
+          {
+            (int64_t)pg_->getRank(),
+            (int64_t)str.length(),
+          }, {torch::kLong})
+      };
+
+      // ProcessGroup is not thread-safe when sending with the same tag, hence
+      // the lock
```

Comment:
So the lock here will end up simply making one send execute at a time -- so the only thing that you're parallelizing is the serialization, right?

Reply:
It's one lock per rank. If this RPC agent communicates with different processes, this would still help, no?

Reply:
This changes the ordering semantics of the messages.

E.g. if you want to send 2 messages to the remote, and you queue both of them one after the other without waiting for completion, the execution order is now undefined. Since there are multiple send threads, these messages can be picked up at the same time, and then those threads race to grab the lock.

If those messages are calls that mutate local state on the remote, this can become messy.

Reply:
> if you want to send 2 messages to the remote, and you queue both of them one after the other without waiting for completion, the execution order is now undefined. 

Will the undefined order be a problem? I thought we are not going to enforce any order on RPC execution? Even if we guarantee FIFO on the caller side, there is still no promise on the order of execution on the callee side, right? Because there might be multiple threads working in parallel serving requests.

> Since there are multiple send threads, these messages can be picked up at the same time, and then those threads race to grab the lock.

Ahh, yes, this will hurt performance.

> If those messages are calls that mutate local state on the remote, this can become messy.

Can this be considered as an application problem? They apps would like to enforce order on mutating some global var, they should add syncs properly?

Reply:
> Can this be considered as an application problem? They apps would like to enforce order on mutating some global var, they should add syncs properly?

Yes, we can approach it like this. But from the user point of view it will be unexpected, I think. If you make two calls one after the other and they get reordered, but only if certain timing conditions hit, it will be a PITA to debug. And we can ask ourselves how many users will 1) read the docs, and 2) apply strict ordering. I think it's worth thinking about some kind of ordered mode, which is an optional thing that forces sequencing of RPCs to a single remote. Then in examples we can refer to it, and it will become clear soon enough that by default you shouldn't expect the call order and execution order to be the same.

Reply:
I don't think we should worry about ordering of RPC calls, most RPC frameworks don't guarantee this. For example, in thrift if you kick off two RPC calls one after the other, there is no guarantee which one will complete first.

@mrshenli Does the thread-safety issue of ProcessGroup apply to all types of ProcessGroups? I was looking at the implementation for ProcessGroupGloo and it seems like the transport layer does have some of its own locking. If send() is not thread-safe for the same tag for ProcessGroupGloo, could you point out which part is not thread-safe?

Reply:
@pritamdamania87 One of the issue is that, the receiver excepts to see a preamble followed by a payload. If we don't have a lock here, the receiver could get two consecutive preambles.

Reply:
@mrshenli: Why don't we send the preamble and the payload together in one send call?